sosw/components/sigv4.py

```diff
@@ -0,0 +1,263 @@
+        Adapted from https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html
+        """
+
+        self.amzdate, self.datestamp = self.get_headers_and_credential_date()
+        aws_sigv4_headers = self.get_aws_sigv4_headers(request)
+        request.headers.update(aws_sigv4_headers)
+
+        return request
+
+
+    def get_aws_sigv4_headers(self, request):
+        """
+        Returns a dictionary containing the necessary headers for Amazon's
+        signature version 4 signing process.
+        """
+
+        parsed_url = urlparse(request.url)
+        logger.debug("Request URL: %s", parsed_url)
+
+        canonical_querystring = self.get_canonical_querystring(parsed_url)
+        canonical_uri = self.get_canonical_uri(parsed_url)
+        canonical_headers = self.get_canonical_headers(parsed_url)
+
+        payload_hash = self.get_payload_hash(request)
+        signed_headers = 'host;x-amz-date;x-amz-security-token' if self.aws_session_token else 'host;x-amz-date'
+
+        canonical_request = '\n'.join([request.method, canonical_uri, canonical_querystring,
+                                       canonical_headers, signed_headers, payload_hash])
+        logger.debug("Canonical Request: '%s'", canonical_request)
+
+        authorization_header = self.get_authorization_header(canonical_request, signed_headers)
+
+        headers = {
+            'Authorization':        authorization_header,
+            'X-AMZ-Date':           self.amzdate,
+            'x-amz-content-sha256': payload_hash,
+        }
+
+        if self.aws_session_token:
+            request.headers['x-amz-security-token'] = self.aws_session_token
+
+        logger.debug("Generated Request Headers: %s", headers)
+
+        return headers
+
+
+    @staticmethod
+    def sign(key, msg):
+        return hmac.new(key, msg.encode('utf-8'), hashlib.sha256).digest()
+
+
+    @staticmethod
+    def get_headers_and_credential_date():
+        """
+        Create a date for headers and the credential string
+        """
+
+        date_utc_now = datetime.datetime.utcnow()
+        amzdate = date_utc_now.strftime('%Y%m%dT%H%M%SZ')
+        datestamp = date_utc_now.strftime('%Y%m%d')
+        logger.debug("Starting authentication with amzdate=%s and datestamp=%s", amzdate, datestamp)
+
+        return amzdate, datestamp
+
+
+    @staticmethod
+    def get_payload_hash(request):
+        """
+        Create payload hash. For GET requests, the payload is an empty string ("")
+        """
+
+        if request.method == 'GET':
+            payload = ''.encode('utf-8')
+
+        else:
+            if request.body:
+                payload = request.body if isinstance(request.body, bytes) else request.body.encode('utf-8')
+
+            else:
+                payload = b''
+
+        logger.debug("Request Body: <bytes> %s", payload)
+
+        return hashlib.sha256(payload).hexdigest()
+
+
+    @staticmethod
+    def get_canonical_querystring(parsed_url):
```

Comment:
Why invent broken bicycles?
``urllib.urlencode``

It can accept ordered list of tuples of params and values you want to encode.